File: src/diagnostic.zig
const std = @import("std");
const fmt = std.fmt;
const mem = std.mem;
const ansi = @import("./ansi.zig");

const ArrayList = std.ArrayList;
const Allocator = std.mem.Allocator;
const Io = std.Io;
const Span = @import("./location.zig").Span;
const TokenType = @import("./lexer/Token.zig").TokenType;
const CowStr = @import("./CowStr.zig").CowStr;

pub const Diagnostic = struct {
    allocator: Allocator,
    absolute_filename: ?CowStr = null,
    source: ?CowStr = null,
    inner: ?DiagnosticInner = null,

    const Self = @This();

    /// This function takes ownerships of CowStrs.
    pub fn initMetadata(
        self: *Self,
        absolute_filename: ?CowStr,
        source: ?CowStr,
    ) void {
        if (absolute_filename) |af| {
            if (self.absolute_filename) |*old_af| {
                old_af.deinit(self.allocator);
                old_af.* = af;
            } else {
                self.absolute_filename = af;
            }
        }
        if (source) |s| {
            if (self.source) |*old_s| {
                old_s.deinit(self.allocator);
                old_s.* = s;
            } else {
                self.source = s;
            }
        }
    }

    pub fn initMetadataAlloc(
        self: *Self,
        absolute_filename: ?[]const u8,
        source: ?[]const u8,
    ) !void {
        if (absolute_filename) |af| {
            const new_af = try CowStr.init(.Owned, .{
                self.allocator,
                af,
            });
            if (self.absolute_filename) |*old_af| {
                old_af.deinit(self.allocator);
                old_af.* = new_af;
            } else {
                self.absolute_filename = new_af;
            }
        }
        errdefer if (self.absolute_filename) |*af| af.deinit(self.allocator);

        if (source) |s| {
            const new_s = try CowStr.init(.Owned, .{
                self.allocator,
                s,
            });
            if (self.source) |*old_s| {
                old_s.deinit(self.allocator);
                old_s.* = new_s;
            } else {
                self.source = new_s;
            }
        }
    }

    pub fn initDiagInner(self: *Self, diag_inner: DiagnosticInner) void {
        if (self.inner) |*old_inner| {
            old_inner.deinit(self.allocator);
            old_inner.* = diag_inner;
        } else {
            self.inner = diag_inner;
        }
    }

    pub fn deinit(self: *Self) void {
        if (self.absolute_filename) |*af| af.deinit(self.allocator);
        if (self.source) |*source| source.deinit(self.allocator);
        if (self.inner) |*inner| inner.deinit(self.allocator);
    }

    pub fn prettyPrint(
        self: Self,
        no_color: bool,
    ) !void {
        if (self.inner) |inner| try inner.prettyPrint(
            self.allocator,
            self.absolute_filename,
            self.source,
            no_color,
        );
    }

    pub fn reset(self: *Self) void {
        self.deinit();
        self.* = .{};
    }
};

pub const DiagnosticKind = enum {
    ParseError,
    IOError,
};

pub const DiagnosticInner = union(DiagnosticKind) {
    ParseError: ParseDiagnostic,
    IOError: IODiagnostic,

    const Self = @This();

    pub fn deinit(self: *Self, allocator: Allocator) void {
        switch (self.*) {
            inline else => |*inner| inner.deinit(allocator),
        }
    }

    fn prettyPrint(
        self: Self,
        allocator: Allocator,
        absolute_filename: ?CowStr,
        source: ?CowStr,
        no_color: bool,
    ) !void {
        switch (self) {
            inline else => |diag| try diag.prettyPrint(
                allocator,
                absolute_filename,
                source,
                no_color,
            ),
        }
    }
};

pub const IODiagnostic = struct {
    msg: ArrayList(u8),
    note_msg: ?ArrayList(u8) = null,
    span: ?Span,

    const Self = @This();

    pub fn init(
        allocator: Allocator,
        span: ?Span,
        comptime fmt_str: []const u8,
        args: anytype,
    ) !Self {
        var io_err_msg = try ArrayList(u8).initCapacity(allocator, 30);
        errdefer io_err_msg.deinit(allocator);

        try io_err_msg.print(allocator, fmt_str, args);

        return Self{ .msg = io_err_msg, .span = span };
    }

    pub fn initWithNote(
        allocator: Allocator,
        span: ?Span,
        comptime fmt_str: []const u8,
        args: anytype,
        comptime fmt_note_str: []const u8,
        note_args: anytype,
    ) !Self {
        var io_err_msg = try ArrayList(u8).initCapacity(allocator, 30);
        errdefer io_err_msg.deinit(allocator);
        var io_note_msg = try ArrayList(u8).initCapacity(allocator, 30);
        errdefer io_note_msg.deinit(allocator);

        try io_err_msg.print(allocator, fmt_str, args);
        try io_note_msg.print(allocator, fmt_note_str, note_args);

        return Self{ .msg = io_err_msg, .note_msg = io_note_msg, .span = span };
    }

    pub fn deinit(self: *Self, allocator: Allocator) void {
        self.msg.deinit(allocator);
        if (self.note_msg) |*note_msg| note_msg.deinit(allocator);
    }

    fn prettyPrint(
        self: Self,
        allocator: Allocator,
        absolute_filename: ?CowStr,
        source: ?CowStr,
        no_color: bool,
    ) !void {
        _ = allocator;
        _ = absolute_filename;
        _ = source;

        const stderr_handle = std.fs.File.stderr();
        var stderr_buf: [4096]u8 = undefined;
        var stderr = stderr_handle.writer(&stderr_buf);

        if (stderr_handle.supportsAnsiEscapeCodes() and !no_color) {
            try stderr.interface.print(
                ansi.@"error" ++ "error: " ++ ansi.reset ++ "{s}\n",
                .{self.msg.items},
            );
            if (self.note_msg) |note_msg| {
                try stderr.interface.print(
                    ansi.note ++ "note: " ++ ansi.reset ++ "{s}\n",
                    .{note_msg.items},
                );
            }
        } else {
            try stderr.interface.print("error: {s}\n", .{self.msg.items});
            if (self.note_msg) |note_msg| {
                try stderr.interface.print("note: {s}\n", .{note_msg.items});
            }
        }

        try stderr.interface.flush();
    }
};

pub const ParseDiagnostic = struct {
    err_info: ParseErrorInfo = .None,
    span: ?Span = null,

    const Self = @This();

    pub const ParseErrKind = enum(u8) {
        None = 0,
        EofErr,
        PremiereErr,
        TokenExpected,
        NameMissErr,
        IsNotOpened,
        IsNotClosed,
        IllegalUseErr,
        Deprecated,
        LuaLabelNotFound,
        DuplicatedPyLabel,
        InvalidPycode,
        DisallowPycode,
        PyEvalFailed,
        NotLocatedInVeryFirst,
        DoubleUsed,
        InvalidLatexEngine,
        VestiInternal,
    };

    pub const ParseErrorInfo = union(ParseErrKind) {
        None,
        EofErr,
        PremiereErr,
        TokenExpected: struct {
            expected: []const TokenType,
            obtained: ?TokenType,
        },
        NameMissErr: TokenType,
        IsNotOpened: struct {
            open: []const TokenType,
            close: TokenType,
        },
        IsNotClosed: struct {
            open: []const TokenType,
            close: TokenType,
        },
        IllegalUseErr: []const u8,
        Deprecated: []const u8,
        LuaLabelNotFound: ArrayList(u8),
        DuplicatedPyLabel: []const u8,
        InvalidPycode,
        DisallowPycode,
        PyEvalFailed: struct {
            err_msg: ArrayList(u8),
            err_detail: ArrayList(u8),
        },
        NotLocatedInVeryFirst: TokenType,
        DoubleUsed: TokenType,
        InvalidLatexEngine: []const u8,
        VestiInternal: []const u8,

        fn deinit(self: *@This(), allocator: Allocator) void {
            switch (self.*) {
                .LuaLabelNotFound => |*label| label.deinit(allocator),
                .PyEvalFailed => |*inner| {
                    inner.err_msg.deinit(allocator);
                    inner.err_detail.deinit(allocator);
                },
                else => {},
            }
        }
    };

    pub fn pyLabelNotFound(
        allocator: Allocator,
        span: Span,
        label_str: []const u8,
    ) !Self {
        var label = try ArrayList(u8).initCapacity(allocator, label_str.len);
        errdefer label.deinit(allocator);
        try label.appendSlice(allocator, label_str);

        return Self{
            .err_info = ParseErrorInfo{ .LuaLabelNotFound = label },
            .span = span,
        };
    }

    pub fn pyEvalFailed(
        allocator: Allocator,
        span: ?Span,
        comptime fmt_str: []const u8,
        args: anytype,
        err_detail_str: [:0]const u8,
    ) !Self {
        var err_msg = try ArrayList(u8).initCapacity(allocator, 30);
        errdefer err_msg.deinit(allocator);
        try err_msg.print(allocator, fmt_str, args);

        var err_detail = try ArrayList(u8).initCapacity(allocator, err_detail_str.len);
        errdefer err_detail.deinit(allocator);
        try err_detail.appendSlice(allocator, err_detail_str);

        return Self{
            .err_info = ParseErrorInfo{ .PyEvalFailed = .{
                .err_msg = err_msg,
                .err_detail = err_detail,
            } },
            .span = span,
        };
    }

    pub fn deinit(self: *Self, allocator: Allocator) void {
        self.err_info.deinit(allocator);
    }

    fn errorMsg(
        self: Self,
        allocator: Allocator,
    ) !ArrayList(u8) {
        var output = try ArrayList(u8).initCapacity(allocator, 50);
        errdefer output.deinit(allocator);
        var aw: Io.Writer.Allocating = .fromArrayList(allocator, &output);
        errdefer aw.deinit();

        switch (self.err_info) {
            .None => try aw.writer.writeAll("<none>"), // TODO: is it the best?
            .EofErr => try aw.writer.print("EOF character was found", .{}),
            .PremiereErr => try aw.writer.print("PremiereErr\n", .{}),
            .TokenExpected => |info| {
                if (info.expected.len == 1) {
                    try aw.writer.print(
                        "{f} was expected but got {?f}",
                        .{ info.expected[0], info.obtained },
                    );
                } else {
                    var tmp = Io.Writer.Allocating.init(allocator);
                    defer tmp.deinit();

                    try tmp.writer.writeByte('[');
                    for (info.expected) |info_inner| {
                        try tmp.writer.print(" {f}", .{info_inner});
                    }
                    try tmp.writer.writeAll(" ]");

                    try aw.writer.print(
                        "{s} was expected but got {?f}",
                        .{ tmp.written(), info.obtained },
                    );
                }
            },
            .NameMissErr => |toktype| try aw.writer.print(
                "{f} should have a name",
                .{toktype},
            ),
            .IsNotOpened => |info| {
                if (info.open.len == 1) {
                    try aw.writer.print(
                        "either {f} was not opened with {f}",
                        .{ info.close, info.open[0] },
                    );
                } else {
                    var tmp = Io.Writer.Allocating.init(allocator);
                    defer tmp.deinit();

                    try tmp.writer.writeByte('[');
                    for (info.open) |info_inner| {
                        try tmp.writer.print(" {f}", .{info_inner});
                    }
                    try tmp.writer.writeAll(" ]");

                    try aw.writer.print(
                        "either {f} was not opened with {s}",
                        .{ info.close, tmp.written() },
                    );
                }
            },
            .IsNotClosed => |info| {
                if (info.open.len == 1) {
                    try aw.writer.print(
                        "{f} was not closed with {f}",
                        .{ info.open[0], info.close },
                    );
                } else {
                    var tmp = Io.Writer.Allocating.init(allocator);
                    defer tmp.deinit();

                    try tmp.writer.writeByte('[');
                    for (info.open) |info_inner| {
                        try tmp.writer.print(" {f}", .{info_inner});
                    }
                    try tmp.writer.writeAll(" ]");

                    try aw.writer.print(
                        "either {s} was not closed with {f}",
                        .{ tmp.written(), info.close },
                    );
                }
            },
            .Deprecated => |info| try aw.writer.print(
                "deprecated token was found. Replace `{s}` instead",
                .{info},
            ),
            inline .IllegalUseErr,
            .VestiInternal,
            => |info| try aw.writer.writeAll(info),
            .InvalidPycode => try aw.writer.writeAll("invalid pycode was found"),
            .DisallowPycode => try aw.writer.writeAll("nested pycode is not allowed"),
            .LuaLabelNotFound => |label| try aw.writer.print(
                "label `{s}` is not found",
                .{label.items},
            ),
            .DuplicatedPyLabel => |label| try aw.writer.print(
                "label `{s}` is duplicated",
                .{label},
            ),
            .PyEvalFailed => |inner| try aw.writer.print(
                "python exception occured: {s}",
                .{inner.err_msg.items},
            ),
            .NotLocatedInVeryFirst => |tok| try aw.writer.print(
                "{f} must located in the very first line of the vesti code",
                .{tok},
            ),
            .DoubleUsed => |tok| try aw.writer.print(
                "{f} must be used only once",
                .{tok},
            ),
            .InvalidLatexEngine => |engine| try aw.writer.print(
                "invalid latex engine name {s} was found",
                .{engine},
            ),
        }

        return aw.toArrayList();
    }

    fn noteMsg(
        self: Self,
        allocator: Allocator,
    ) !?ArrayList(u8) {
        return switch (self.err_info) {
            .LuaLabelNotFound => blk: {
                var output = try ArrayList(u8).initCapacity(allocator, 50);
                errdefer output.deinit(allocator);

                try output.print(
                    allocator,
                    "labels should be declared before it is used",
                    .{},
                );
                break :blk output;
            },
            .PyEvalFailed => |inner| blk: {
                var output = try ArrayList(u8).initCapacity(allocator, 50);
                errdefer output.deinit(allocator);

                try output.print(allocator, "{s}", .{inner.err_detail.items});
                break :blk output;
            },
            .InvalidLatexEngine => blk: {
                var output = try ArrayList(u8).initCapacity(allocator, 50);
                errdefer output.deinit(allocator);

                try output.print(
                    allocator,
                    "valid ones are `plain`, `pdf`, `xe`, and `lua`",
                    .{},
                );
                break :blk output;
            },
            else => null,
        };
    }

    fn prettyPrint(
        self: Self,
        allocator: Allocator,
        absolute_filename: ?CowStr,
        source: ?CowStr,
        no_color: bool,
    ) !void {
        std.debug.assert(source != null);

        const stderr_handle = std.fs.File.stderr();
        var stderr_buf: [4096]u8 = undefined;
        var stderr = stderr_handle.writer(&stderr_buf);

        const filename = if (absolute_filename) |af| blk: {
            const current_dir = try std.fs.cwd().realpathAlloc(allocator, ".");
            defer allocator.free(current_dir);
            break :blk try std.fs.path.relative(
                allocator,
                current_dir,
                af.toStr(),
            );
        } else "^.^";
        defer if (absolute_filename != null) allocator.free(filename);

        var err_msg = try self.errorMsg(allocator);
        defer err_msg.deinit(allocator);

        if (self.span) |span| {
            var line_iter = mem.splitScalar(u8, source.?.toStr(), '\n');

            for (1..span.start.row) |_| _ = line_iter.next();
            const line = line_iter.next() orelse @panic("span.start.row is invalid");

            const source_trim = mem.trimRight(u8, line, "\r");
            const underline = if (span.start.row == span.end.row) blk: {
                @branchHint(.likely);
                break :blk try allocator.alloc(u8, span.end.col - 1);
            } else blk: {
                break :blk try allocator.alloc(u8, source_trim.len);
            };
            defer allocator.free(underline);

            @memset(underline[0 .. span.start.col - 1], ' ');
            if (span.start.row == span.end.row) {
                @branchHint(.likely);
                @memset(underline[span.start.col - 1 .. span.end.col - 1], '^');
            } else {
                @memset(underline[span.start.col - 1 ..], '^');
            }

            if (stderr_handle.supportsAnsiEscapeCodes() and !no_color) {
                try stderr.interface.print(
                    ansi.makeAnsi(null, .Bold) ++ "{s}:{}:{}: " ++ ansi.@"error" ++
                        "error: " ++ ansi.reset ++ "{s}\n" ++
                        "    {s}\n" ++
                        ansi.makeAnsi(.BrightGreen, null) ++ "    {s}\n" ++ ansi.reset,
                    .{
                        filename,      span.start.row, span.start.col,
                        err_msg.items, source_trim,    underline,
                    },
                );

                var note_msg = try self.noteMsg(allocator);
                if (note_msg) |*nm| {
                    defer nm.deinit(allocator);
                    try stderr.interface.print(
                        ansi.note ++ "note: " ++ ansi.reset ++ "{s}\n",
                        .{nm.items},
                    );
                }
            } else {
                try stderr.interface.print(
                    \\{s}:{}:{}: error: {s}
                    \\    {s}
                    \\    {s}
                    \\
                , .{
                    filename,      span.start.row, span.start.col,
                    err_msg.items, source_trim,    underline,
                });
                var note_msg = try self.noteMsg(allocator);
                if (note_msg) |*nm| {
                    defer nm.deinit(allocator);
                    try stderr.interface.print("note: {s}\n", .{nm.items});
                }
            }
        } else {
            if (stderr_handle.supportsAnsiEscapeCodes() and !no_color) {
                try stderr.interface.print(
                    ansi.makeAnsi(null, .Bold) ++ "{s}: " ++ ansi.@"error" ++
                        "error: " ++ ansi.reset ++ "{s}\n" ++ ansi.reset,
                    .{ filename, err_msg.items },
                );

                var note_msg = try self.noteMsg(allocator);
                if (note_msg) |*nm| {
                    defer nm.deinit(allocator);
                    try stderr.interface.print(
                        ansi.note ++ "note: " ++ ansi.reset ++ "{s}\n",
                        .{nm.items},
                    );
                }
            } else {
                try stderr.interface.print(
                    \\{s}: error: {s}
                    \\
                , .{ filename, err_msg.items });

                var note_msg = try self.noteMsg(allocator);
                if (note_msg) |*nm| {
                    defer nm.deinit(allocator);
                    try stderr.interface.print("note: {s}\n", .{nm.items});
                }
            }
        }

        try stderr.interface.flush();
    }
};

File: src/diagnostic.rs
use std::borrow::Cow;
use std::fmt::Write as _;

use crate::lexer::token::TokenType;
use crate::location::Span;

// TODO: Replace this into crossterm one
// Minimal ANSI helpers
mod ansi {
    pub const RESET: &str = "\x1b[0m";
    pub const ERROR: &str = "\x1b[1;31m";
    pub const NOTE: &str = "\x1b[1;36m";
    pub const BOLD: &str = "\x1b[1m";
    pub const BRIGHT_GREEN: &str = "\x1b[92m";
}

#[derive(Default)]
pub struct Diagnostic<'s> {
    pub absolute_filename: Option<Cow<'s, str>>,
    pub source: Option<Cow<'s, str>>,
    pub inner: Option<DiagnosticInner<'s>>,
    /// When set, `main` must not pretty-print the diagnostic again (the
    /// compiler already printed it).
    pub lock_print_at_main: bool,
}

impl<'s> Diagnostic<'s> {
    pub fn new() -> Self {
        Diagnostic::default()
    }

    pub fn init_metadata(
        &mut self,
        absolute_filename: Option<Cow<'s, str>>,
        source: Option<Cow<'s, str>>,
    ) {
        if let Some(af) = absolute_filename {
            self.absolute_filename = Some(af);
        }
        if let Some(s) = source {
            self.source = Some(s);
        }
    }

    pub fn init_diag_inner(&mut self, diag_inner: DiagnosticInner<'s>) {
        self.inner = Some(diag_inner);
    }

    pub fn set_parse_error(&mut self, err_info: ParseErrorInfo<'s>, span: Option<Span>) {
        self.init_diag_inner(DiagnosticInner::ParseError(ParseDiagnostic {
            err_info,
            span,
        }));
    }

    pub fn reset(&mut self) {
        *self = Diagnostic::default();
    }

    /// Render the diagnostic into a string (used both by `pretty_print` and by
    /// tests). Returns `None` when there is no inner diagnostic.
    pub fn render(&self, no_color: bool) -> Option<String> {
        let inner = self.inner.as_ref()?;
        let mut out = String::new();
        inner.render(
            self.absolute_filename.as_deref(),
            self.source.as_deref(),
            no_color,
            &mut out,
        );
        Some(out)
    }

    pub fn pretty_print(&self, no_color: bool) {
        if let Some(text) = self.render(no_color) {
            eprint!("{text}");
        }
    }
}

pub enum DiagnosticInner<'s> {
    ParseError(ParseDiagnostic<'s>),
    IoError(IoDiagnostic),
}

impl<'s> DiagnosticInner<'s> {
    fn render(
        &self,
        absolute_filename: Option<&str>,
        source: Option<&str>,
        no_color: bool,
        out: &mut String,
    ) {
        match self {
            DiagnosticInner::ParseError(d) => d.render(absolute_filename, source, no_color, out),
            DiagnosticInner::IoError(d) => d.render(no_color, out),
        }
    }
}

pub struct IoDiagnostic {
    pub msg: String,
    pub note_msg: Option<String>,
    pub span: Option<Span>,
}

impl IoDiagnostic {
    pub fn new(span: Option<Span>, msg: String) -> Self {
        IoDiagnostic {
            msg,
            note_msg: None,
            span,
        }
    }

    pub fn with_note(span: Option<Span>, msg: String, note: String) -> Self {
        IoDiagnostic {
            msg,
            note_msg: Some(note),
            span,
        }
    }

    fn render(&self, no_color: bool, out: &mut String) {
        if !no_color {
            let _ = write!(out, "{}error: {}{}\n", ansi::ERROR, ansi::RESET, self.msg);
            if let Some(note) = &self.note_msg {
                let _ = write!(out, "{}note: {}{}\n", ansi::NOTE, ansi::RESET, note);
            }
        } else {
            let _ = write!(out, "error: {}\n", self.msg);
            if let Some(note) = &self.note_msg {
                let _ = write!(out, "note: {}\n", note);
            }
        }
    }
}

#[derive(Default)]
pub struct ParseDiagnostic<'s> {
    pub err_info: ParseErrorInfo<'s>,
    pub span: Option<Span>,
}

#[derive(Default)]
pub enum ParseErrorInfo<'s> {
    #[default]
    None,
    BegenvUnderflow,
    ChangeEngineTwice,
    DefunParamOverflow(usize),
    Deprecated(&'static str),
    DisallowLuacode,
    DuplicatedLuaLabel(String),
    EnvInsideDefun,
    EofErr,
    IllegalUseErr(&'static str),
    InvalidBuiltin(String),
    InvalidDefunKind(String),
    InvalidDefunParam(usize),
    InvalidLatexEngine(&'s str),
    InvalidLuacode,
    InvalidTokenFound,
    IsNotClosed {
        open: &'static [TokenType<'static>],
        close: TokenType<'static>,
    },
    IsNotOpened {
        open: &'static [TokenType<'static>],
        close: TokenType<'static>,
    },
    LuaEvalFailed {
        err_msg: String,
        err_detail: String,
    },
    LuaLabelNotFound(String),
    MathmodeInMath,
    ModuleNotFound(String),
    NameMissErr(TokenType<'static>),
    PreambleErr,
    TextmodeInText,
    TokenExpected {
        expected: &'static [TokenType<'static>],
        obtained: Option<TokenType<'s>>,
    },
    TooManyBegenv(u8),
    VestiInternal(&'static str),
    WrongBuiltin {
        name: Cow<'s, str>,
        note: &'static str,
    },
}

impl<'s> ParseDiagnostic<'s> {
    pub fn lua_label_not_found(span: Span, label: &str) -> Self {
        ParseDiagnostic {
            err_info: ParseErrorInfo::LuaLabelNotFound(label.to_owned()),
            span: Some(span),
        }
    }

    pub fn lua_eval_failed(span: Option<Span>, msg: String, detail: &str) -> Self {
        ParseDiagnostic {
            err_info: ParseErrorInfo::LuaEvalFailed {
                err_msg: msg,
                err_detail: detail.to_owned(),
            },
            span,
        }
    }

    fn error_msg(&self) -> String {
        let mut out = String::new();
        match &self.err_info {
            ParseErrorInfo::None => out.push_str("<none>"),
            ParseErrorInfo::EofErr => out.push_str("end of file was detected"),
            ParseErrorInfo::PreambleErr => out.push_str("PremiereErr\n"),
            ParseErrorInfo::TokenExpected { expected, obtained } => {
                if expected.len() == 1 {
                    let _ = write!(
                        out,
                        "{} was expected but got {}",
                        expected[0],
                        opt_tok(obtained.as_ref())
                    );
                } else {
                    let mut tmp = String::from("[");
                    for e in expected.iter() {
                        let _ = write!(tmp, " {e}");
                    }
                    tmp.push_str(" ]");
                    let _ = write!(
                        out,
                        "{tmp} was expected but got {}",
                        opt_tok(obtained.as_ref())
                    );
                }
            }
            ParseErrorInfo::NameMissErr(toktype) => {
                let _ = write!(out, "{toktype} should have a name");
            }
            ParseErrorInfo::IsNotOpened { open, close } => {
                if open.len() == 1 {
                    let _ = write!(out, "either {close} was not opened with {}", open[0]);
                } else {
                    let mut tmp = String::from("[");
                    for o in open.iter() {
                        let _ = write!(tmp, " {o}");
                    }
                    tmp.push_str(" ]");
                    let _ = write!(out, "either {close} was not opened with {tmp}");
                }
            }
            ParseErrorInfo::IsNotClosed { open, close } => {
                if open.len() == 1 {
                    let _ = write!(out, "{} was not closed with {close}", open[0]);
                } else {
                    let mut tmp = String::from("[");
                    for o in open.iter() {
                        let _ = write!(tmp, " {o}");
                    }
                    tmp.push_str(" ]");
                    let _ = write!(out, "either {tmp} was not closed with {close}");
                }
            }
            ParseErrorInfo::TextmodeInText => out.push_str("`#textmode` found in text"),
            ParseErrorInfo::MathmodeInMath => out.push_str("`#mathmode` found in math"),
            ParseErrorInfo::TooManyBegenv(_) => out.push_str("too many `begenv` was found"),
            ParseErrorInfo::BegenvUnderflow => out.push_str("there is no `begenv` to match"),
            ParseErrorInfo::InvalidTokenFound => out.push_str("invalid token was found"),
            ParseErrorInfo::Deprecated(info) => {
                let _ = write!(out, "deprecated token was found. Replace `{info}` instead");
            }
            ParseErrorInfo::ModuleNotFound(mod_path) => {
                let _ = write!(out, "There is no module at {mod_path}");
            }
            ParseErrorInfo::InvalidDefunKind(defun_kind) => {
                let _ = write!(out, "String `{defun_kind}` is not a valid defun attribute");
            }
            ParseErrorInfo::DefunParamOverflow(val) => {
                let _ = write!(
                    out,
                    "2 to the power of {val} exceeds max value of 2^{}",
                    std::mem::size_of::<usize>()
                );
            }
            ParseErrorInfo::InvalidBuiltin(builtin_fnt) => {
                let _ = write!(out, "builtin `#{builtin_fnt}` is not defined");
            }
            ParseErrorInfo::WrongBuiltin { name, .. } => {
                let _ = write!(out, "wrong usage for builtin `#{name}`");
            }
            ParseErrorInfo::InvalidDefunParam(val) => {
                let _ = write!(
                    out,
                    "parameter number `{val}` is wrong one for a function parameter"
                );
            }
            ParseErrorInfo::EnvInsideDefun => {
                out.push_str("`useenv` cannot be used inside `defun` body")
            }
            ParseErrorInfo::IllegalUseErr(info) | ParseErrorInfo::VestiInternal(info) => {
                out.push_str(info)
            }
            ParseErrorInfo::InvalidLuacode => out.push_str("invalid lua code was found"),
            ParseErrorInfo::DisallowLuacode => out.push_str("nested lua code is not allowed"),
            ParseErrorInfo::LuaLabelNotFound(label) => {
                let _ = write!(out, "label `{label}` is not found");
            }
            ParseErrorInfo::DuplicatedLuaLabel(label) => {
                let _ = write!(out, "label `{label}` is duplicated");
            }
            ParseErrorInfo::LuaEvalFailed { err_msg, .. } => {
                let _ = write!(out, "lua exception occured: {err_msg}");
            }
            ParseErrorInfo::ChangeEngineTwice => {
                out.push_str("engine type is tried to changed in twice, which is not allowed")
            }
            ParseErrorInfo::InvalidLatexEngine(engine) => {
                let _ = write!(out, "invalid latex engine name {engine} was found");
            }
        }
        out
    }

    fn note_msg(&self) -> Option<String> {
        match &self.err_info {
            ParseErrorInfo::EofErr => Some(
                "usually, this error occurs when the brace does not match.\n\
                 maybe the chatacter pointed has no pair in the code."
                    .to_owned(),
            ),
            ParseErrorInfo::TooManyBegenv(count) => {
                Some(format!("maximum allowed `begenv` is {count}."))
            }
            ParseErrorInfo::LuaLabelNotFound(_) => {
                Some("labels should be declared before it is used".to_owned())
            }
            ParseErrorInfo::LuaEvalFailed { err_detail, .. } => Some(err_detail.clone()),
            ParseErrorInfo::InvalidLatexEngine(_) => {
                Some("valid ones are `plain`, `pdf`, `xe`, and `lua`".to_owned())
            }
            ParseErrorInfo::WrongBuiltin { note, .. } => Some((*note).to_owned()),
            ParseErrorInfo::InvalidDefunParam(val) => Some(format!(
                "because of the LaTeX and Vesti internal, only values not divisible by 10 are valid for a function parameter. But you gave `{val}`."
            )),
            ParseErrorInfo::EnvInsideDefun => Some(
                "Internally vesti uses \\def family to define latex functions.\n\
                 However, latex hates to use environment inside of \\def.\n\
                 For this reason, vesti emits an error to notice to change."
                    .to_owned(),
            ),
            ParseErrorInfo::ChangeEngineTwice => Some(
                "There are two ways to change latex engine: in `first.lua` and `compty`.\n\
                 Vesti allows to use either of them, not both"
                    .to_owned(),
            ),
            ParseErrorInfo::VestiInternal(_) => Some(
                "Unexpected behavior was found.\n\
                 If this error is raised, please make an issue on github.\n\
                 addr: https://github.com/e0328eric/vesti"
                    .to_owned(),
            ),
            _ => None,
        }
    }

    fn render(
        &self,
        absolute_filename: Option<&str>,
        source: Option<&str>,
        no_color: bool,
        out: &mut String,
    ) {
        let source = source.expect("source must be set for ParseDiagnostic");
        let filename = absolute_filename.unwrap_or("^.^");
        let err_msg = self.error_msg();

        if let Some(span) = self.span {
            let line = source
                .split('\n')
                .nth(span.start.row() - 1)
                .expect("span.start.row is invalid");
            let source_trim = line.trim_end_matches('\r');

            // Build the caret underline.
            let mut underline = String::new();
            for _ in 0..span.start.column().saturating_sub(1) {
                underline.push(' ');
            }
            if span.start.row() == span.end.row() {
                let count = span.end.column().saturating_sub(span.start.column());
                for _ in 0..count {
                    underline.push('^');
                }
            } else {
                let count = source_trim
                    .chars()
                    .count()
                    .saturating_sub(span.start.column().saturating_sub(1));
                for _ in 0..count {
                    underline.push('^');
                }
            }

            if !no_color {
                let _ = write!(
                    out,
                    "{}{}:{}:{}: {}error: {}{}\n    {}\n{}    {}\n{}",
                    ansi::BOLD,
                    filename,
                    span.start.row(),
                    span.start.column(),
                    ansi::ERROR,
                    ansi::RESET,
                    err_msg,
                    source_trim,
                    ansi::BRIGHT_GREEN,
                    underline,
                    ansi::RESET,
                );
                if let Some(nm) = self.note_msg() {
                    let _ = writeln!(out, "{}note: {}{}", ansi::NOTE, ansi::RESET, nm);
                }
            } else {
                let _ = write!(
                    out,
                    "{}:{}:{}: error: {}\n    {}\n    {}\n",
                    filename,
                    span.start.row(),
                    span.start.column(),
                    err_msg,
                    source_trim,
                    underline,
                );
                if let Some(nm) = self.note_msg() {
                    let _ = writeln!(out, "note: {}", nm);
                }
            }
        } else if !no_color {
            let _ = write!(
                out,
                "{}{}: {}error: {}{}\n{}",
                ansi::BOLD,
                filename,
                ansi::ERROR,
                ansi::RESET,
                err_msg,
                ansi::RESET,
            );
            if let Some(nm) = self.note_msg() {
                let _ = writeln!(out, "{}note: {}{}", ansi::NOTE, ansi::RESET, nm);
            }
        } else {
            let _ = writeln!(out, "{}: error: {}", filename, err_msg);
            if let Some(nm) = self.note_msg() {
                let _ = writeln!(out, "note: {}", nm);
            }
        }
    }
}

fn opt_tok(t: Option<&TokenType<'_>>) -> String {
    match t {
        Some(t) => format!("{t}"),
        None => "null".to_owned(),
    }
}
